pyxrf/model/setting.py

```diff
@@ -211,34 +335,111 @@ def get_roi_sum(self):
-                logger.debug('Calculation is done for {}, {}, {}'.format(v.prefix,
-                                                                         fname, k))
-            roi_result[v.prefix+'_'+fname] = temp
-            return roi_result
-
-
-def calculate_roi(data3D, e_linear, e_offset, range_v):
+
+        datav = self.data_sets[self.data_title]
+
+        logger.info(f"Computing ROIs for dataset {self.data_title} ...")
+
+        temp = {}
+        data_raw = np.asarray(datav.raw_data)
+
+        if self.subtract_background:
+
+            logger.info(f"Subtracting background ...")
+
+            num_processors_to_use = multiprocessing.cpu_count()
+            logger.info(f"Cpu count: {format(num_processors_to_use)}")
+            pool = multiprocessing.Pool(num_processors_to_use)
+
+            result_pool = [pool.apply_async(
+                _subtract_background_one_line,
+                (data_raw[n, :, :],
+                 self.parameters['e_offset']['value'],
+                 self.parameters['e_linear']['value'],
+                 self.parameters['e_quadratic']['value'],
+                 self.parameters['non_fitting_values']['background_width']))
+                 for n in range(data_raw.shape[0])]
+
+            data_roi = []
+            for r in result_pool:
+                data_roi.append(r.get())
+
+            pool.terminate()
+            pool.join()
+
+            data_roi = np.array(data_roi)
+
+            logger.info(f"Background subtraction completed.")
+
+        else:
+            data_roi = data_raw
+
+        for k, v in six.iteritems(self.roi_dict):
+            leftv = v.left_val/1000
+            rightv = v.right_val/1000
+            sum2D = calculate_roi(data_roi,
+                                  e_offset=self.parameters['e_offset']['value'],
+                                  e_linear=self.parameters['e_linear']['value'],
+                                  range_v=[leftv, rightv])
+            temp.update({k: sum2D})
+            logger.debug(f"Calculation is completed for {v.prefix}, {self.data_title}, {k}")
+
+        # Save ROI data to HDF5 file
+        self.saveROImap_to_hdf(temp)
+
+        # Add scalers to the ROI dataset, so that they can be selected from Image Wizard.
+        # We don't want to save scalers to the file, since they are already in the file.
+        # So we add scalers after data is saved.
+        scaler_key = f"{self.data_title_base}_scaler"
+        if scaler_key in self.img_dict:
+            temp.update(self.img_dict[scaler_key])
+
+        roi_result[f"{self.data_title_adjusted}_roi"] = temp
+
+        logger.info("ROI is computed.")
+        return roi_result
+
+    def saveROImap_to_hdf(self, data_dict_roi):
+
+        if 'det1' in self.data_title:
+            inner_path = 'xrfmap/det1'
+        elif 'det2' in self.data_title:
+            inner_path = 'xrfmap/det2'
+        elif 'det3' in self.data_title:
+            inner_path = 'xrfmap/det3'
```

Comment:
It probably can be done in a more elegant way, such as:
```py
inner_path = None
for det in ['det1', 'det2', 'det3']:
    if det in self.data_title:
        inner_path = f'xrfmap/{det}'
        break
if inner_path is None:
    inner_path = 'xrfmap/detsum'
```
This way one can easily add new detectors in _one_ place.

Reply:
Fixed